File: packages/web/lib/analytics.ts
'use client'

/**
 * PostHog product analytics — the single, centralized event surface for the web app
 * (Epic 11.6). Every event name lives in ANALYTICS_EVENTS; nothing else in the app
 * should pass a raw string to posthog.capture().
 *
 * Privacy: we identify by Clerk user id and attach only coarse, non-sensitive
 * properties (tier, school, graduation, work auth, job portal, fit category). NEVER
 * pass resume content, answer/message text, or personal contact info through here.
 *
 * No-op safety: if NEXT_PUBLIC_POSTHOG_KEY is unset (local dev) PostHog is never
 * initialized, so every helper below short-circuits and analytics is a silent no-op.
 * All calls are wrapped so a telemetry hiccup can never break a user flow.
 */
import posthog from 'posthog-js'

import { readConsent } from '@/lib/cookie-consent'

export const ANALYTICS_EVENTS = {
  // ---- Funnel (signup → activation → paid) ----
  // signed_up fires SERVER-SIDE from the Clerk webhook (source of truth for account
  // creation); listed here only so the catalog is complete.
  SIGNED_UP: 'signed_up',
  ONBOARDING_COMPLETED: 'onboarding_completed',
  RESUME_UPLOADED: 'resume_uploaded',
  RESUME_SCORED: 'resume_scored',
  JOBS_VIEWED: 'jobs_viewed',
  SCOUT_RUN_STARTED: 'scout_run_started',
  // application_completed fires SERVER-SIDE from the apply task (accuracy-critical).
  APPLICATION_COMPLETED: 'application_completed',
  UPGRADE_VIEWED: 'upgrade_viewed',
  CHECKOUT_STARTED: 'checkout_started',
  PLAN_CHANGE_STARTED: 'plan_change_started',
  // subscription_activated fires SERVER-SIDE from the Stripe webhook (payment source
  // of truth); listed here only for catalog completeness.
  SUBSCRIPTION_ACTIVATED: 'subscription_activated',

  // ---- Feature usage ----
  COPILOT_MESSAGE_SENT: 'copilot_message_sent',
  RESUME_REWRITE_USED: 'resume_rewrite_used',
  RESUME_TAILORED_FOR_JOB: 'resume_tailored_for_job',
  APPLICATION_MANUALLY_MANAGED: 'application_manually_managed',
  NOTIFICATION_CLICKED: 'notification_clicked',
  BILLING_PORTAL_OPENED: 'billing_portal_opened',

  // ---- Mascot tutorial / profile nudge ----
  INTRO_TOUR_SHOWN: 'intro_tour_shown',
  INTRO_TOUR_CONTINUED: 'intro_tour_continued',
  INTRO_TOUR_SKIPPED: 'intro_tour_skipped',
  PROFILE_NUDGE_SHOWN: 'profile_nudge_shown',
  PROFILE_NUDGE_DISMISSED: 'profile_nudge_dismissed',
  PROFILE_NUDGE_CTA_CLICKED: 'profile_nudge_cta_clicked',
  PAGE_INTRO_SHOWN: 'page_intro_shown',
  PAGE_INTRO_DISMISSED: 'page_intro_dismissed',
  FIRST_SCOUT_RUN_ACK_SHOWN: 'first_scout_run_ack_shown',
  FIRST_SCOUT_RUN_ACK_DISMISSED: 'first_scout_run_ack_dismissed',
} as const

export type AnalyticsEvent =
  (typeof ANALYTICS_EVENTS)[keyof typeof ANALYTICS_EVENTS]

/** Non-sensitive properties allowed on events. Free-form, but keep keys snake_case
 * and never include resume/answer/message/contact content. */
export type AnalyticsProps = Record<
  string,
  string | number | boolean | null | undefined
>

const POSTHOG_KEY = process.env.NEXT_PUBLIC_POSTHOG_KEY
const POSTHOG_HOST =
  process.env.NEXT_PUBLIC_POSTHOG_HOST || 'https://us.i.posthog.com'

let initialized = false
// Set once the user revokes consent in-session: PostHog may already be initialized
// (can't be un-init'd), so we hard-stop every helper as well as opting out.
let consentRevoked = false

/** True only when a PostHog key is configured AND the user has accepted cookies.
 * Used to gate every call: local dev (no key) and pre-/post-consent never touch the
 * SDK. Consent is read from the first-party cookie on each call. */
export function isAnalyticsEnabled(): boolean {
  return Boolean(POSTHOG_KEY) && !consentRevoked && readConsent() === 'accepted'
}

/** Initialize the PostHog browser SDK exactly once. Safe to call repeatedly and on
 * the server (it no-ops unless a key is set and `window` exists). Does NOTHING until
 * the user has accepted cookies — this deferral is the real consent gate, since
 * posthog.init() is what captures the first pageview and sets PostHog cookies. */
export function initAnalytics(): void {
  if (!POSTHOG_KEY || typeof window === 'undefined') return
  if (readConsent() !== 'accepted') return
  if (initialized) {
    // Already initialized this session. If the user had revoked and is now re-
    // accepting, opt back into capturing — init() won't run a second time.
    if (consentRevoked) {
      consentRevoked = false
      try {
        posthog.opt_in_capturing()
      } catch {
        /* no-op */
      }
    }
    return
  }
  consentRevoked = false
  posthog.init(POSTHOG_KEY, {
    api_host: POSTHOG_HOST,
    // Only create person profiles for users we explicitly identify (logged-in) —
    // keeps anonymous marketing traffic out of person-level analytics.
    person_profiles: 'identified_only',
    // Capture pageviews (incl. SPA route changes) so funnel steps like "explore
    // viewed" are available, but DO NOT autocapture clicks/inputs: autocapture can
    // scrape element/DOM text, and Scout's DOM contains resume/answer content.
    capture_pageview: true,
    capture_pageleave: true,
    autocapture: false,
    disable_session_recording: true,
  })
  initialized = true
}

/** Identify the current user by their Clerk id and merge in coarse person traits.
 * Call on login and whenever traits change. */
export function identifyUser(distinctId: string, traits?: AnalyticsProps): void {
  if (!isAnalyticsEnabled() || !distinctId) return
  try {
    posthog.identify(distinctId, traits)
  } catch {
    /* analytics must never break the app */
  }
}

/** Register super properties attached to EVERY subsequent event (e.g. tier), so we
 * don't have to thread them through individual capture calls. */
export function registerSuperProperties(props: AnalyticsProps): void {
  if (!isAnalyticsEnabled()) return
  try {
    posthog.register(props)
  } catch {
    /* no-op */
  }
}

/** Reset identity on logout so the next user isn't merged into the previous person. */
export function resetUser(): void {
  if (!isAnalyticsEnabled()) return
  try {
    posthog.reset()
  } catch {
    /* no-op */
  }
}

/** Stop analytics when the user revokes consent. If PostHog was already initialized
 * earlier in the session it can't be un-init'd, so we opt out of capturing (stops all
 * network + clears stored data) and reset identity. The consentRevoked flag then makes
 * every helper a no-op for the rest of the session; a fresh page load won't re-init
 * because initAnalytics() re-checks the (now 'rejected') consent cookie. */
export function disableAnalytics(): void {
  consentRevoked = true
  if (!initialized) return
  try {
    posthog.opt_out_capturing()
    posthog.reset()
  } catch {
    /* no-op */
  }
}

/** Capture a product event. The ONLY way the app emits analytics events. */
export function track(event: AnalyticsEvent, properties?: AnalyticsProps): void {
  if (!isAnalyticsEnabled()) return
  try {
    posthog.capture(event, properties)
  } catch {
    /* no-op */
  }
}

File: packages/api/core/analytics.py
"""PostHog product analytics — server-side capture (Epic 11.6).

Used for the events that must be reliable and can't be trusted to the browser:
- subscription_activated  — fired from the Stripe webhook (payment source of truth).
- application_completed    — fired from the apply task, only when an application truly
                             reaches "applied" status (never from optimistic UI).
- signed_up                — fired from the Clerk webhook (account-creation source of
                             truth).

Design mirrors core/observability.py (Sentry):
- NO-OP when POSTHOG_KEY is unset (local dev): the client is never created and every
  helper short-circuits, so nothing crashes, blocks, or spams.
- distinct_id is ALWAYS the Clerk user id, so server events tie to the same person the
  browser SDK identifies — funnels stitch across client and server.
- Privacy: never pass resume / answer / message / personal-contact content. Identify by
  id + coarse properties (tier, portal, etc.) only.
"""

import logging
import os

from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

# Canonical server-emitted event names (snake_case). These mirror the names in the
# frontend catalog (packages/web/lib/analytics.ts) so client+server events line up in
# the same PostHog funnel.
EVENT_SIGNED_UP = "signed_up"
EVENT_APPLICATION_COMPLETED = "application_completed"
EVENT_SUBSCRIPTION_ACTIVATED = "subscription_activated"

_DEFAULT_HOST = "https://us.i.posthog.com"

_client = None
_initialized = False


def _get_client():
    """Lazily build the PostHog client once. Returns None (disabled) when no key is
    configured or init fails — callers treat None as a silent no-op."""
    global _client, _initialized
    if _initialized:
        return _client

    _initialized = True
    key = (os.getenv("POSTHOG_KEY") or "").strip()
    if not key:
        logger.info("PostHog disabled (POSTHOG_KEY unset)")
        _client = None
        return None

    try:
        from posthog import Posthog

        _client = Posthog(
            project_api_key=key,
            host=(os.getenv("POSTHOG_HOST") or _DEFAULT_HOST).strip(),
            # The request originates from our server, not the user — don't geo-locate
            # events by the server's IP.
            disable_geoip=True,
        )
        logger.info("PostHog initialized (server-side)")
    except Exception as exc:  # never let analytics setup break startup
        logger.warning("PostHog init failed: %s", exc)
        _client = None

    return _client


def capture(distinct_id, event, properties=None, *, flush=False):
    """Capture a server-side event for the given Clerk user id.

    No-ops when PostHog is unconfigured or distinct_id is missing. Never raises —
    analytics must not break a webhook or a Celery task. Pass flush=True for
    payment/accuracy-critical events so they're delivered promptly even if the process
    is about to go idle.
    """
    client = _get_client()
    if client is None or not distinct_id:
        return
    try:
        client.capture(event, distinct_id=distinct_id, properties=properties or {})
        if flush:
            client.flush()
    except Exception as exc:
        logger.warning("PostHog capture failed (event=%s): %s", event, exc)


def set_user_properties(distinct_id, properties):
    """Update a user's PostHog person properties (e.g. subscription_plan on upgrade)
    without emitting a counted event. Keeps tier-based segmentation accurate the moment
    a plan changes server-side. No-ops when unconfigured."""
    client = _get_client()
    if client is None or not distinct_id or not properties:
        return
    try:
        client.set(distinct_id=distinct_id, properties=properties)
        client.flush()
    except Exception as exc:
        logger.warning("PostHog set properties failed: %s", exc)
